Remove Frame class from DefaultHttp2RemoteFlowController. Fixes #3488

buchgr opened the pull request:
Motivation:

For every write of a flow controlled frame (HEADERS, DATA) we are allocating
a Frame object that is not necessary anymore as it does not maintain any
state, besides the payload.

Modifications:

Remove the Frame class and directly add the payload to the pending write queue.

Result:

One few object allocation per write of a flow controlled frame.

buchgr commented:
@nmittler @Scottmitch please review.

normanmaurer commented:
@buchgr you not asked for me to review but did it anyway, hope you not mind =P

buchgr commented:
@normanmaurer sorry I can't accept that ;-)

nmittler commented:
@buchgr Not sure if you rebased ... I had to click on the commit to review @normanmaurer's comments.

Overall changes LGTM!

Scottmitch commented:
@buchgr - Looks good!  I followed the trend and also just left comments on the commit.

Scottmitch commented:
Link to https://github.com/netty/netty/issues/3488

buchgr commented:
I believe I addressed all the comments, except that one https://github.com/buchgr/netty/commit/4f31b649607da75b85fbef7de2376b811c2aae67#commitcomment-10155708

@Scottmitch @normanmaurer I don't understand what the issue is with that? You want to switch the semantics to increment and pass in a negative value? If so, why is that better?

Scottmitch commented:
@buchgr - I think @normanmaurer was suggesting storing the result of the negated value in a temporary variable because it is done 2 times in the method.  I am suggesting taking it a step back and looking at how this method is used with another method (`decrementPendingBytes`) that also negates the value.  Consider just negate the value 1 time and pass to functions....or maybe just call `incrementPendingBytes` directly with the negated value?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java:
```diff
@@ -335,8 +334,8 @@ private static void writeChildNode(FlowState state) {
      */
-    final class FlowState {
-        private final Deque<Frame> pendingWriteQueue;
+    private class FlowState {
```
add `final`?

buchgr replied:
I removed it because, what's the point of a private final class? Nobody can extend it anyway.

Scottmitch replied:
A class within the `DefaultHttp2RemoteFlowController` scope could :)  Not that is it too hard to derive but it may provide clarity into the hierarchy intentions/assumptions and also a hint to JIT.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java:
```diff
@@ -350,7 +349,7 @@ private static void writeChildNode(FlowState state) {
         FlowState(Http2Stream stream, int initialWindowSize) {
```
I think the package private access *maybe* for 2 reasons:

1. Used in tests?
2. Avoid synthetic method generation?

So I'm not sure we need the package private access but if we could make more restrictive that is good too.

buchgr replied:
@Scottmitch hmm well it's not used in tests and IMHO we should not care about synthetic methods at all, unless I am missing something. WDYT?

buchgr commented:
@Scottmitch hmm well it's not used in tests and IMHO we should not care about synthetic methods at all, unless I am missing something. WDYT?

buchgr commented:
@Scottmitch @normanmaurer TBH I think the code is very expressive and good the way it is. I can change it to whatever you suggest, but I don't see a need for this. Also, I just moved the code. It's in Netty like this right now. Didn't change a line :-).

Scottmitch commented:
> IMHO we should not care about synthetic methods at all

I think we should try to avoid synthetic methods if they can easily be avoided.  If the access type is made too restrictive (i.e. `private`, ...) then additional static methods will be synthesized to allow the outer class to invoke them.

> I think the code is very expressive and good the way it is

I agree.  However I also think that using a temporary to store the negative value won't negatively impact expressiveness.  My `take a step back` suggestion was a more "invasive" optimization that we can pass on if you think this will perturb things too much.

> Also, I just moved the code. It's in Netty like this right now. Didn't change a line :-).

You are doing a great job and if there are small/localized optimizations/improvements you can make along the way that is even better :)

buchgr commented:
@Scottmitch 

> However I also think that using a temporary to store the negative value won't negatively impact expressiveness.

but do you think it improves it? :-) I don't TBH.

normanmaurer commented:
@buchgr not sure if it improves it but I kind of always try to not trust the JIT too much ;)

Scottmitch commented:
@buchgr - I stand by my previous statement :)  I do think there is some clarity/efficiency gains rather than continuously negating the value.

buchgr commented:
I am fine with it, I added a local variable - it's not something we should spend much time on :-). But yeah it does optimize this http://en.wikipedia.org/wiki/Common_subexpression_elimination

So we good? Ready for Merge? :-) Did I miss anything?

Scottmitch commented:
@buchgr - Thanks.  Can you just add the `final` keyword back and we can cherry pick?

nmittler commented:
cherrypicked into 4.1 and master.

Scottmitch commented:
@buchgr -  Great work...less object allocations for the win!
